site/src/sphinx/advanced-retry.rst

```diff
@@ -1,8 +1,225 @@
+
+RetryingClient_
+---------------
+
+You can just use ``decorator`` method in ClientBuilder_ to build a RetryingHttpClient_:
+
+.. code-block:: java
+
+    import com.linecorp.armeria.client.ClientBuilder;
+    import com.linecorp.armeria.client.HttpClient;
+    import com.linecorp.armeria.common.HttpRequest;
+    import com.linecorp.armeria.common.HttpResponse;
+    import com.linecorp.armeria.client.retry.RetryingHttpClient;
+    import com.linecorp.armeria.client.retry.RetryStrategy;
+
+    RetryStrategy strategy = RetryStrategy.onServerErrorStatus();
+    HttpClient client = new ClientBuilder(...)
+                              .decorator(HttpRequest.class, HttpResponse.class,
+                                         RetryingHttpClient.newDecorator(strategy))
+                              .build(HttpClient.class);
+
+    client.execute(...).aggregate().join();
+
+or even simply,
+
+.. code-block:: java
+
+    import com.linecorp.armeria.client.HttpClientBuilder;
+
+    RetryStrategy strategy = RetryStrategy.onServerErrorStatus();
+    HttpClient client = new HttpClientBuilder(...)
+                          .decorator(RetryingHttpClient.newDecorator(strategy))
+                          .build();
+
+    client.execute(...).aggregate().join();
+
+That is it. The client will keep attempting until it succeeds or the number of attempts exceeds the default
+number of max attempts. You can configure the ``defaultMaxAttempts`` when making the decorator using
+``RetryingHttpClient.newDecorator(strategy, defaultMaxAttempts)``. Meanwhile, the ``strategy`` will decide to
+retry depending on the response. In this case, the client retries when it receives ``5xx`` response error.
+
+RetryStrategy
+-------------
+
+You can customize the ``strategy`` by implementing RetryStrategy_.
+
+.. code-block:: java
+
+    import com.linecorp.armeria.client.retry.Backoff;
+    import com.linecorp.armeria.common.HttpStatus;
+
+    new RetryStrategy<HttpRequest, HttpResponse>() {
+        final Backoff backoff = RetryStrategy.defaultBackoff;
+
+        @Override
+        public CompletableFuture<Optional<Backoff>> shouldRetry(HttpRequest request,
+                                                                HttpResponse response) {
+            return response.aggregate().handle((result, cause) -> { // don't do get() or join()!
+                if (cause != null) {
+                    if (cause instanceof ResponseTimeoutException) {
+                        return Optional.of(backoff);
+                    }
+                } else if (result.headers().status() == HttpStatus.CONFLICT) {
+                    return Optional.of(backoff);
+                }
+                return Optional.empty(); // return no backoff not to retry anymore
+            });
+        }
+    };
+
+This will retry when the response's status is ``409``.
+
+.. note::
+
+    We declare a ``backoff`` as a member and return it. RetryingClient_ internally tracks the
```

Comment:
- return it -> reuse it when a strategy returns it, so that we do not return a different Backoff_ instance for each shouldRetry().